src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -84,5 +224,187 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+
+            PackageSpec currentPackageSpec = await GetPackageSpecAsync(token);
+            PackageSpec cachedPackageSpec = null;
+
+            if (_lastPackageSpec != null)
+            {
+                _lastPackageSpec.TryGetTarget(out cachedPackageSpec);
+            }
+
+            string assetsFilePath = await GetAssetsFilePathAsync();
+            var assets = new FileInfo(assetsFilePath);
+
+            bool cacheMissAssets = (assets.Exists && assets.LastWriteTimeUtc > _lastTimeAssetsModified);
+
+            if (cacheMissAssets || (this is CpsPackageReferenceProject && !ReferenceEquals(currentPackageSpec, cachedPackageSpec)))
+            {
+                _lastTimeAssetsModified = assets.LastWriteTimeUtc;
+                _lastPackageSpec = new WeakReference<PackageSpec>(currentPackageSpec);
+                IsInstalledAndTransitiveComputationNeeded = true;
+            }
+
+            return (currentPackageSpec, assetsFilePath);
+        }
+
+        /// <summary>
+        /// Obtains targets section from project assets file (project.assets.json)
+        /// </summary>
+        /// <param name="ct">Cancellation token for async operation</param>
+        /// <returns>A list of dependencies, indexed by framework/RID</returns>
+        /// <remarks>Assets file reading occurs in a background thread</remarks>
+        /// <seealso cref="GetAssetsFilePathAsync"/>
+        private async ValueTask<IList<LockFileTarget>> GetTargetsListAsync(string assetsFilePath, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            await TaskScheduler.Default;
+
+            LockFile lockFile = LockFileUtilities.GetLockFile(assetsFilePath, NullLogger.Instance);
+
+            return lockFile?.Targets;
+        }
+
+        /// <summary>
+        /// Runs Depth First Search recursively to mark current and dependend nodes with top dependency
+        /// </summary>
+        /// <param name="top">Top, Direct dependency</param>
+        /// <param name="current">Current package/node to visit</param>
+        /// <param name="graph">Package dependency graph, from assets file</param>
+        /// <param name="visited">Dictionary to remember visited nodes</param>
+        /// <param name="fxRidEntry">Framework/Runtime-ID associated with current <paramref name="graph"/></param>
+        private void MarkTransitiveOrigin(PackageReference top, PackageReference current, LockFileTarget graph, HashSet<PackageIdentity> visited, FrameworkRuntimePair fxRidEntry, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+
+            LockFileTargetLibrary node = default;
+
+            // Find first target node that matches current
+            foreach (LockFileTargetLibrary lib in graph.Libraries)
+            {
+                if (lib.Type == LibraryType.Package.Value
+                    && string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase))
```

Comment:
Does package version matters here?

Reply:
Can be removed. I will just keep it just in case.